### ISSUE TITLE
InvalidCastException when observing a List

### ISSUE THREAD

**Issue by Lythom** [nadako/TinkStateSharp#34]
Hey there, I got an exception, not sure if it's a misuse or if there is an actual bug but I isolated the error in a test case :

The idea is that an autorun subscription will raise an InvalidCastException in this situation.

i'm guessing this is because I'm updating the collection inside of the Auto() which obviously is an antipattern. There is no exception if a remove the _dailyStats "reset" part of the code and manually insert values before observing. Maybe there is a way to get a detection of the "write while subscribing" case and get a more meaninglful exception message?

```csharp
    [Test]
    public void GetDailyStats_WithAutoRun_ShouldNotCrash() {
        ObservableList<float> _dailyStats = Observable.List<float>();
        Observable<IReadOnlyList<float>> DailyStats() => _dailyStats.Observe();

        float GetDailyStats(int dailIdx) {
            _dailyStats.Clear();
            while (_dailyStats.Count <= dailIdx) _dailyStats.Add(0);
            return DailyStats().Value[dailIdx];
        }

        // Arrange
        var autoRunState = Observable.Auto(() => {
            var dailyStatsValue1 = GetDailyStats(1);
            return dailyStatsValue1;
        });

        // Act & Assert
        Assert.DoesNotThrow(() => {
            var value = autoRunState.Value; // This should not throw
            _dailyStats[1] = 5;
            value = autoRunState.Value; // This should not throw either
            Assert.AreEqual(5, value);
        });
    }
```

cast exception:
```cs
Exception of type 'System.InvalidCastException' was thrown: Unable to cast object of type 'TinkState.Internal.Subscription`1[[TinkState.Internal.ObservableList`1[[System.Single, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], Nadako.TinkState, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]]' to type 'TinkState.Internal.Subscription`1[[System.Collections.Generic.IReadOnlyList`1[[System.Single, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]'
```

Exception stack:
```cs
Expected: No Exception to be thrown
  But was:  <System.InvalidCastException: Specified cast is not valid.
  at TinkState.Internal.AutoObservable`1[T].SubscribeTo[R] (TinkState.Internal.DispatchingObservable`1[T] source) [0x00051] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:330 
  at TinkState.Internal.AutoObservable.Track[T] (TinkState.Internal.DispatchingObservable`1[T] o) [0x00016] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:24 
  at TinkState.Internal.ObservableList`1[T].TinkState.Observable<System.Collections.Generic.IReadOnlyList<T>>.get_Value () [0x00000] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\ObservableList_ObservableOfReadOnlyList.cs:10 
  at PersistantStateTests+<>c__DisplayClass11_0.<GetDailyStats_WithAutoRun_ShouldNotCrash>g__GetDailyStats|1 (System.Int32 dailIdx) [0x00035] in D:\lonestone\MomentumManitou\Assets\Tests\PersistantStateTests.cs:182 
  at PersistantStateTests+<>c__DisplayClass11_0.<GetDailyStats_WithAutoRun_ShouldNotCrash>b__2 () [0x00001] in D:\lonestone\MomentumManitou\Assets\Tests\PersistantStateTests.cs:187 
  at TinkState.Internal.SyncComputation`1[T].GetNext () [0x00001] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\Computation.cs:24 
  at TinkState.Internal.AutoObservable.ComputeFor[R] (TinkState.Internal.Derived o, TinkState.Internal.Computation`1[TResult] computation) [0x0000d] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:45 
  at TinkState.Internal.AutoObservable`1[T].DoCompute () [0x00029] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:203 
  at TinkState.Internal.AutoObservable`1[T].GetCurrentValue (System.Boolean force) [0x0003e] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:156 
  at TinkState.Internal.AutoObservable`1[T].GetRevision () [0x00023] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:287 
  at TinkState.Internal.Subscription`1[T]..ctor (TinkState.Internal.DispatchingObservable`1[T] source) [0x0000f] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:394 
  at TinkState.Internal.AutoObservable`1[T].SubscribeTo[R] (TinkState.Internal.DispatchingObservable`1[T] source) [0x00017] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:321 
  at TinkState.Internal.AutoObservable.Track[T] (TinkState.Internal.DispatchingObservable`1[T] o) [0x00016] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:24 
  at TinkState.Internal.AutoObservable`1[T].get_Value () [0x00000] in .\Library\PackageCache\me.nadako.tinkstate\TinkState\Runtime\Internal\AutoObservable.cs:134 
  at PersistantStateTests+<>c__DisplayClass11_0.<GetDailyStats_WithAutoRun_ShouldNotCrash>b__3 () [0x00001] in D:\lonestone\MomentumManitou\Assets\Tests\PersistantStateTests.cs:193 
  at NUnit.Framework.Constraints.ThrowsConstraint+VoidInvocationDescriptor.Invoke () [0x00000] in <f577cd9686154d6ab10efd7e65e64366>:0 
  at NUnit.Framework.Constraints.ThrowsConstraint+ExceptionInterceptor.Intercept (System.Object invocation) [0x00041] in <f577cd9686154d6ab10efd7e65e64366>:0 >

Stacktrace: 
at PersistantStateTests.GetDailyStats_WithAutoRun_ShouldNotCrash () [0x0002c] in D:\lonestone\MomentumManitou\Assets\Tests\PersistantStateTests.cs:192
```

![image](https://github.com/user-attachments/assets/cb4ba983-9109-4f91-b2ca-f6319694e9b6)

**Comment by nadako** [comment]
Hey, thanks for the report. This definitely looks like a legit bug caused by that reckless cast, I'm gonna look into it to see how can we avoid it.

And no, surprisingly it's not caused by changing while calculating (although that's definitely an antipattern!), rather it's caused by using both `ObservableList` and its `Observable<IReadOnlyList>` aspect (the same object under the hood) in the same auto-observable computation, messing up internal subscription object casting.

You can quickfix your specifc code here by just using `_dailyStats` instead of `DailyStats().Value`, but of course I assume in the real code it's not that simple :)

Here's a stripped down version explaining what's happening:
```csharp
using TinkState;

class Playground
{
	static void Main()
	{
		var list = Observable.List<int>();
		var observedList = list.Observe();

		list.Add(1);

		var auto = Observable.Auto(() =>
		{
			// creates a subscription to ObservableList<T>
			Use(list.Count);

			// tries to reuse the subscription for the same object
			// but fails to cast it for IReadOnlyList<T>
			Use(observedList.Value);

			return 1;
		});

		Use(auto.Value);
	}

	static void Use<T>(T v) {}
}
```

**Comment by Lythom** [comment]
Ahh I was not sure _dailyStats would register a use in Auto methods and that's why I took the version with "Observe()". Now I understand it did registered the use properly haha.

Anyway it's good I got a closer look at that code to fix that anti-pattern, the list reset now happens during a proper lifecycle timing outside auto.
I think I'll keep the internal use of the `Observable<IReadOnlyList>` (which is the only possible public access for the rest of the code), so that whatever the code in the Auto method it will always register an `Observable<IReadOnlyList>` and never a `ObservableList`.

Thanks for the quick answer and for the clarification!
